unstructured/utils.py

```diff
@@ -280,3 +281,392 @@ def scarf_analytics():
+        ngram = []
+        for j in range(n):
+            ngram.append(s[i + j])
+        ngrams_list.append(tuple(ngram))
+    return ngrams_list
+
+
+def calculate_shared_ngram_percentage(
+    first_string: str,
+    second_string: str,
+    n: int,
+) -> (float, List):
+    """Calculate the percentage of common_ngrams between string_A and string_B
+    with reference to the total number of ngrams in string_A"""
+
+    first_string_ngrams = ngrams(first_string.split(), n)
+    second_string_ngrams = ngrams(second_string.split(), n)
+
+    if not first_string_ngrams:
+        return 0
+
+    common_ngrams = set(first_string_ngrams) & set(second_string_ngrams)
+    percentage = (len(common_ngrams) / len(first_string_ngrams)) * 100
+    return percentage, common_ngrams
+
+
+def calculate_largest_ngram_percentage(first_string: str, second_string: str) -> (float, List, str):
+    """Iteratively calculate_shared_ngram_percentage starting from the biggest
+    ngram possible until is >0.0%"""
+
+    shared_ngrams = []
+    if len(first_string.split()) < len(second_string.split()):
+        n = len(first_string.split()) - 1
+    else:
+        n = len(second_string.split()) - 1
+        first_string, second_string = second_string, first_string
+    ngram_percentage = 0
+    while not ngram_percentage:
+        ngram_percentage, shared_ngrams = calculate_shared_ngram_percentage(
+            first_string,
+            second_string,
+            n,
+        )
+        if n == 0:
+            break
+        else:
+            n -= 1
+    return round(ngram_percentage, 2), shared_ngrams, str(n)
+
+
+def is_parent_box(
+    parent_target: Union[List, Tuple],
+    child_target: Union[List, Tuple],
+    add: float = 0.0,
+) -> bool:
+    """True if the child_target bounding box is nested in the parent_target.
+    Box format: [x_bottom_left, y_bottom_left, x_top_right, y_top_right].
+    The parameter 'add' is the pixel error tolerance for extra pixels outside the parent region"""
+
+    if len(parent_target) != 4:
+        return False
+
+    if add and len(parent_target) == 4:
+        parent_target = list(parent_target)
+        parent_target[0] -= add
+        parent_target[1] -= add
+        parent_target[2] += add
+        parent_target[3] += add
+
+    if (
+        len(child_target) == 4
+        and (child_target[0] >= parent_target[0] and child_target[1] >= parent_target[1])
+        and (child_target[2] <= parent_target[2] and child_target[3] <= parent_target[3])
+    ):
+        return True
+    if len(child_target) == 2 and (
+        parent_target[0] <= child_target[0] <= parent_target[2]
+        and parent_target[1] <= child_target[1] <= parent_target[3]
+    ):
+        return True
+    return False
+
+
+def calculate_overlap_percentage(
+    box1: Union[List, Tuple],
+    box2: Union[List, Tuple],
+    intersection_ratio_method: str = "total",
+):
+    """Box format: [x_bottom_left, y_bottom_left, x_top_right, y_top_right].
```

Comment:
Would be difficult to change the function to work with our current format (x_topleft,y_topleft, x_bottom_right,y_bottom_right)?

Reply:
This is working with the coordinate format present in the element list output. No necessary transform. Just pass the list of elements exactly how the partition method is giving them to you. For instance: 
```
'coordinates': {'points': ((374.0989074707031,
     257.9061279296875),
    (374.0989074707031, 282.63995361328125),
    (613.169189453125, 282.63995361328125),
    (613.169189453125, 257.9061279296875)),
```

Inside the methods I work with the `bottom_left` and `top_right` corners. This methods are only going to be called when `catch_overlapping_and_nested_bboxes` is called. So, changing to the other corners won't make any difference in practice... I am not sure what you mean with our `current format` if what we have a an input to `catch_overlapping_and_nested_bboxes` is a list of elements with the coords as I showed. The method `calculate_overlap_percentage` is NOT intended to be run outside `catch_overlapping_and_nested_bboxes`, so I don't see the point. Let me know if you consider the changes are necessary.